llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,386 @@
+#include "llvm/SYCLLowerIR/PropagateAspectUsage.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/DiagnosticPrinter.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+using namespace llvm;
+
+namespace {
+
+class SYCLPropagateAspectUsageLegacyPass : public ModulePass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+  SYCLPropagateAspectUsageLegacyPass() : ModulePass(ID) {
+    initializeSYCLPropagateAspectUsageLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  PropagateAspectUsagePass Impl;
+};
+
+} // namespace
+
+char SYCLPropagateAspectUsageLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLPropagateAspectUsageLegacyPass, "PropagateAspectUsage",
+                "Propagates aspect usage", false, false)
+
+ModulePass *llvm::createPropagateAspectUsagePass() {
+  return new SYCLPropagateAspectUsageLegacyPass();
+}
+
+namespace {
+
+using AspectsSetTy = SmallSet<int, 4>;
+using TypeToAspectsMapTy = std::unordered_map<Type *, AspectsSetTy>;
+
+/// Retrieves from metadata (intel_types_that_use_aspects) types
+/// and aspects these types depend on.
+TypeToAspectsMapTy GetTypesThatUseAspectsFromMetadata(Module &M) {
+  NamedMDNode *Node = M.getNamedMetadata("intel_types_that_use_aspects");
+  TypeToAspectsMapTy Result;
+  if (!Node)
+    return Result;
+
+  LLVMContext &C = M.getContext();
+  for (auto OperandIt : Node->operands()) {
+    MDNode &N = *OperandIt;
+    assert(N.getNumOperands() > 1 && "intel_types_that_use_aspect metadata "
+                                     "shouldn't contain empty metadata nodes");
+
+    MDString *TypeName = cast<MDString>(N.getOperand(0));
+    Type *T = StructType::getTypeByName(C, TypeName->getString());
+    assert(T &&
+           "invalid type referenced by intel_types_that_use_aspect metadata");
+
+    AspectsSetTy &Aspects = Result[T];
+    for (size_t I = 1; I != N.getNumOperands(); ++I) {
+      ConstantAsMetadata *CAM = dyn_cast<ConstantAsMetadata>(N.getOperand(I));
+      assert(CAM && "constant metadata is expected in "
+                    "intel_types_that_use_aspect list's entry");
```

Comment:
Yes, but here we also have the message. Wouldn't it be helpful?

Reply:
I think that the code around is pretty self-explanatory

Reply:
The better point would be: `dyn_cast` performs a runtime check and can return `nullptr`. We do not check that `CAM` is `nullptr` at runtime and therefore we shouldn't be using `dyn_cast`. If we replace `dyn_cast` with `cast`, then that extra assertion won't be needed, because we will never reach it because there is a similar assertion within `cast`.